Create CheckPredicate for performing checks via facts in DL.

markww opened the pull request:
Until now, we've been translating Arcs checks into Datalog rules. While
this gives us a great deal of expressive power, it is tricky to reason
about and requires recompiling before analyzing any Arcs system.

This PR takes the first step in a different direction: instead,
represent checks as regular old facts, with the predicates being
expressed as an ADT. These predicates can then be evaluated inside
Datalog through rules that interpret the predicates encoded in the ADT
rather than encoding the predicates directly as rule bodies.

This PR leaves the infrastructure for the previous way of handling
checks and predicates intact. However, this solution could (and quite
probably should) be used to switch off of that method entirely, allowing
us to precompile the Datalog analysis.

This issue makes progress towards, but does not fix, issue #262 .

bgogul commented on build_defs/test_helpers.bzl:
```diff
@@ -82,3 +89,68 @@ def extracted_datalog_string_test(
+      outs = ["checkNames"],
+      srcs = [check_file],
+      testonly = True,
+      cmd = "cut --delimiter=';' --fields=1 $< > $@",
```
Document what is happening here?

markww replied:
Sounds good, will do.

markww replied:
Done.

bgogul commented on build_defs/test_helpers.bzl:
```diff
@@ -82,3 +89,68 @@ def extracted_datalog_string_test(
+        "--facts=`dirname $(location {})`".format(input_file) for input_file in input_files
+    ]
+
+    native.genrule(
```
A single line comment for each genrule would be useful as documentation.

markww replied:
Will do.

markww replied:
Done.

bgogul commented on src/analysis/souffle/check_predicate.dl:
```diff
@@ -0,0 +1,82 @@
+
+.type Bool <: number
+
+.type CheckPredicate =
```
Is my understanding correct that this is more expressive than what we have in C++ representation?

For instance, the C++ representation can only express something like `ap is tag1 or (tag2 and !tag3)`. However, this can represent `ap1 is tag1 or ap2 is tag2 and ap2 is tag3`. That is, we can combine predicates of different access paths. Is that correct?

markww replied:
No, this is exactly as expressive as what we have in the C++ representation. Each of the ADT branches correspond directly to a `Predicate` we use within checks on the C++ side.

You might be thinking of `claim`s, where we have retreated from the full expressiveness of Arcs because we're not sure what some of that expressiveness means in the overapproximated world of our analysis (for instance, if we claim that some node has `tag1 or tag2`, what does that mean for downstream nodes.

bgogul replied:
The root `CP_TagPresence` also has an `AccessPath` as an argument. So, "(ap1 is tag1) or (ap2 is tag2)` can be represented as follows: 
```
CP_Or { 
   lhs: CP_TagPresence("ap1", "someone", "tag1"),
   rhs: CP_TagPresence("ap2", "someone", "tag2")
}
```

no? On the other hand, the C++ `TagPresence` only has a tag: 
 https://github.com/google-research/raksha/blob/ba14ca7d5a4aebbe31405bc39164dbd6079221e1/src/ir/predicate.h#L232

markww replied:
Oh, I guess you are right. It is more expressive than the original in being able to speak about different `AccessPath`s in the same predicate. I hadn't noticed that.

Another interesting note: because it explicitly lists the `owner` predicate, it can talk about different `owner`s within the same expression tree as well. `owner`s cannot be explicitly listed in the current C++ representation. Currently, the C++ representation translates to a Datalog rule which accepts any owner of the given `tag`; I was thinking of adding a `TagPresenceDynamicOwner` to capture this behavior. I think it is valuable to have a way to require a particular owner as well, hence why I'm thinking of adding an alternative version of `TagPresence`.

bgogul replied:
Sounds good. This is a good change in terms of expressiveness as long as it doesn't blow up the performance. :) Based on your other comment, it seems like we would be OK.

bgogul commented on src/analysis/souffle/check_predicate.dl:
```diff
@@ -0,0 +1,82 @@
+  | CP_And { lhs: CheckPredicate, rhs: CheckPredicate }
+  | CP_Or { lhs: CheckPredicate, rhs: CheckPredicate }
+  | CP_Implies { antecedent: CheckPredicate, consequent: CheckPredicate }
+  | CP_Not { to_negate: CheckPredicate }
```
Consider `CP_Not { some_predicate }`. If `some_predicate` is `true`, will  `CP_Not { some_predicate }` be true or false?

markww replied:
The result of `CP_Not` will be `false` if its input is `true`.

bgogul replied:
Can you document it? It is not clear from the `to_negate` name.

markww replied:
Done.

bgogul commented on src/analysis/souffle/check_predicate.dl:
```diff
@@ -0,0 +1,82 @@
+  checkSubExprs($CP_Implies(ant, cons)), predicateEval(ant, antResult),
+  predicateEval(cons, consResult).
+
+.decl checkAndResult(name: symbol, result: symbol)
+.output checkAndResult(delimiter=";")
+
+checkAndResult(name, "PASS") :-
+  checkAndExpectation(name, pred, _), predicateEval(pred, result), result != 0.
+
+checkAndResult(name, "FAIL") :-
+  checkAndExpectation(name, pred, _), predicateEval(pred, result), result = 0.
+
+// A check which matches the given expectation.
+.decl checkMatchesExpectation(name: symbol)
+.output checkMatchesExpectation(delimiter=";")
+
+checkMatchesExpectation(name) :- checkAndResult(name, result), checkAndExpectation(name, _, result).
```
Should these be part of a test util rather than a part of the core analysis?

markww replied:
Perhaps. Let me see if I can refactor to remove test mechanism from the implementation.

markww replied:
I've refactored the code. It's not quite what you asked for, as the separation between test utils and regular analysis that we have previously accomplished is sort of an artifact of the approach of generating custom datalog which is compiled per-test.   However, I have made the core analysis code not depend upon test relations and have separated the test relations into a conditionally-compiled, test only portion of the Datalog file.

bgogul commented on src/analysis/souffle/check_predicate.dl:
```diff
@@ -0,0 +1,82 @@
+#include "tags.dl"
+
+.type Bool <: number
+
```
Do we have a good sense of what the performance of ADT in datalog is?

markww replied:
That is an interesting question for which I cannot find a full answer online. But I can find some clues.

See this commit to Souffle: https://github.com/souffle-lang/souffle/commit/ddca2fdaeef03a3f6de77abff3a86403fe649b58 .

This commit, in its own words, optimizes ADTs by:

```
encode branch with a single argument as [branch_id, argument]
instead of [branch_id, [argument]]
```

What this tells me is that the various fields of a branch of an ADT are represented as a linked list internally. This means that exploding them and placing each field in a relation, as we do to evaluate the ADT, is probably very cheap. What might not be cheap is getting the last field on a branch of an ADT with many fields, as traversing to each successive field incurs an additional pointer dereference. None of our branches, however, have many fields, so this is probably ok.

That's all I know about ADTs in general. However, I think I have a pretty good understanding of how expensive using ADTs for this particular purpose will be. The evaluation of `check` ADTs will occur at the very end of the analysis; it must, because the leaf `TagPresence` branch consults `!mayHaveTag`, introducing a stratifiability requirement. This means that evaluation of the `CheckPredicate` ADTs will not be interleaved with pushing around tags for the analysis. I believe that this means that this will not occur as part of the inner loop of the analysis, and thus will be of least concern from a performance perspective.

bgogul reviewed (commented):
I like the overall direction, but left some questions and comments. PTAL.

aferr commented on src/analysis/souffle/check_predicate.dl:
```diff
@@ -0,0 +1,82 @@
+.type Bool <: number
+
+.type CheckPredicate =
+  CP_TagPresence { ap: AccessPath, principal: Principal, tag: Tag }
```
Just FYI:
- the error messages when using ADTs can be a bit cryptic at times (you might have already run into this :)
- ADTs are not a usual Datalog feature, they're Souffle-specific. So if we pivot to rolling our own solver, the ADTs might be less portable to our own solver.

You shouldn't necessarily let these points scare you off from using ADTs, but I just thought I'd let you know.

markww replied:
Yes, that's true. If we switched to another Datalog dialect with lists and not ADTs, we could definitely just represent the ADTs as lists with a little less compiler support. If we switched to one without even that, we might be forced to do something ugly by using IDs to point across relations. But I think it is possible with plain ol' Datalog, so I'm not too worried about it for now.

bgogul reviewed (approved):
Looks good modulo refactoring, documentation and clean up suggestions.

markww commented:
@bgogul , I think I addressed all of your comments. If you have a moment, though, it could be good for you to take a look at how I separated out the test relations for `CheckPredicate`s from the core relations. Separating them out into a separate test header included only by the test doesn't make sense in a world where the tests don't contain any custom Datalog, so I separated them out into a conditionally-compiled, test-only region of the `check_predicate.dl` Datalog file.

bgogul commented on src/analysis/souffle/check_predicate.dl:
```diff
@@ -0,0 +1,90 @@
+
+// Relations used to support tests involving check facts. Not present in
+// production builds.
+#ifdef TEST
```
I actually meant, move this to a separate file. OK to leave it as is for now.
